nixos/modules/services/web-servers/h2o/default.nix

```diff
@@ -201,14 +328,32 @@ in
     };
 
     systemd.services.h2o = {
-      description = "H2O web server service";
+      description = "H2O HTTP server";
       wantedBy = [ "multi-user.target" ];
-      after = [ "network.target" ];
+      wants = lib.concatLists (map (certName: [ "acme-finished-${certName}.target" ]) certNames.all);
+      # Since H2O will be hosting the challenges, H2O must be started
+      before = builtins.map (certName: "acme-${certName}.service") certNames.dependent;
+      after =
+        [ "network.target" ]
+        ++ builtins.map (certName: "acme-selfsigned-${certName}.service") certNames.all
+        ++ builtins.map (certName: "acme-${certName}.service") certNames.independent; # avoid loading self-signed key w/ real cert, or vice-versa
 
       serviceConfig = {
-        ExecReload = "${pkgs.coreutils}/bin/kill -HUP $MAINPID";
+        ExecStart = ''${h2oExe} ${
+          lib.escapeShellArgs (
+            lib.optionals (cfg.mode != null) [
+              "--mode"
+              cfg.mode
```

Comment:
Is it useful (or even correct) to let the user select mode? The modes I see in `h2o --help`:

- `worker`: Is this an implementation detail of h2o? Does it make sense to start a worker without a master/daemon orchestrating it?
- `daemon`: I don't think this would play nicely with how the systemd unit is configured (Type=simple, IIRC). See [Example 6. Traditional forking services](https://www.freedesktop.org/software/systemd/man/latest/systemd.service.html#id-1.10.5).
- `master`: This feels like the only mode that would play nicely with this systemd unit configuration.
- `test`: I can't see this being useful here.

Reply:
Daemon feels old school to me. With systemd, services shouldn't need to daemonize themselves.

I see how you're using `test` mode throughout, that looks pretty good to me.

I think we should remove the `mode` option. It seems like a useless footgun. (We can add it back in the future if we're misunderstanding things.)